### ISSUE TITLE
Add a layout capabilitie to {% bootstrap_form %}

### ISSUE THREAD

**Issue by onysos** [dyve/django-bootstrap3#296]
# situation

for now, the tag `{% bootstrap_form %}`  create an horizontal OR inline form.
 
```django
    <form role="form" method="post">
        {% csrf_token %}
        {% bootstrap_form form %}
        {% buttons submit='OK' reset="Cancel" %}{% endbuttons %}
    </form>
```

![forms_orig](https://cloud.githubusercontent.com/assets/3406733/12452191/793d60aa-bf8d-11e5-96e9-398305f2bfc6.png)

# cons
But what if we want two field on the same line for readability purpose ? we then must stop use `{% bootstrap_form %}` and switch to a  `{% bootstrap_field %}` for each fields and create our line with .row and .col-md-6.
```django
    <form role="form" method="post">
        {% csrf_token %}
      <div class="row">
        <div class="col-md-6">
          {%  bootstrap_field form.subject %}
        </div>
        <div class="col-md-6">
          {%  bootstrap_field form.password %}
        </div>
      </div>
      {%  bootstrap_field form.message %}
      {% buttons submit='OK' reset="Cancel" %}{% endbuttons %}
    </form>
```
![form_password_n_subject](https://cloud.githubusercontent.com/assets/3406733/12452317/232b59aa-bf8e-11e5-8194-3b478644db36.png)

this is a probleme as soon as the form is a little dynamic and when some field can be missing (ie because of user perms). so you will soon need to add some `{% if %}` that should not be used if `{% bootstrap_form %}` was used
```django
    <form role="form" method="post">
        {% csrf_token %}
      <div class="row">
        <div class="col-md-6">
          {%  bootstrap_field form.subject %}
        </div>
        <div class="col-md-6">
          {%  bootstrap_field form.password %}
        </div>
      </div>
      {%  bootstrap_field form.message %}
      {%  if form.maybemissing %}
        {%  bootstrap_field form.maybemissing %}
      {%  endif %}
      {% buttons submit='OK' reset="Cancel" %}{% endbuttons %}
    </form>
```
if you don't use `{% if %}` you will have the `BootstrapError: Parameter "field" should contain a valid Django BoundField.`

# enhancement

it seem that the layout of `{% bootstrap_form %}` should be more powerfull and allow us to create a real layout, not just «horizontal» or «inline». 

so, we can have in form.py
```python

class MyLayoutForm(forms.Form):
    layout = [
        ("subject", "password"),
        "message",
        "maybemissing"
    ]
    subject = forms.CharField(
        max_length=100,
        help_text='my_help_text',
        required=True,
        widget=forms.TextInput(attrs={'placeholder': 'placeholdertest'}),
    )
    password = forms.CharField(widget=forms.PasswordInput)
    message = forms.CharField(required=False, help_text='<i>my_help_text</i>')
    maybemissing = forms.EmailField(
        label='Sender © unicode',
        help_text='E.g., "me@example.com"')

    def __init__(self, *args, **kwargs):
        super(MyLayoutForm, self).__init__(*args, **kwargs)
        if whatever:
            self.fields.pop("maybemissing")
```
and keep the simple `{% bootstrap_form %}` in our template, as desired : 

```django
    <form role="form" method="post">
        {% csrf_token %}
        {% bootstrap_form form %}
        {% buttons submit='OK' reset="Cancel" %}{% endbuttons %}
    </form>
```
by adding the "layout" attribute of the Form, the `{% bootstrap_form %}` will use it for the form rendering. if none layout is given, it will create one with all fields, which will render exactly the same way as before. 

the `layout` will allow to create the .row and .col-md-x recursively.
It will take 2 possible type

## list and tuple

so, to get the following form, we can just add a list with tuples : 
```python
    class MyLayoutForm(forms.Form):
        layout = [
            ("subject", "password"),
            "message",
            "maybemissing"
        ]
        ...
```

![form_password_n_subject](https://cloud.githubusercontent.com/assets/3406733/12452317/232b59aa-bf8e-11e5-8194-3b478644db36.png)

## Col and Row

the col and Row are the final step of the layout processing. all given layout will be computed in Col and Row. the `{% bootstrap_form %}` will use it to render the correct layout.

### The Col class

`Col(unicode: fieldname, unicode: class_="col-md-x", *rows)`

the Col represent a `<div class="col-md-xx">`. it can contains other rows or a field of the form (by his name)

```python
Col("message")
```
```python
Col("message", class_="col-md-6 red")
```
```python
Col(
    Row("message"),
    Row("password")
)
```

### the Row class

`Row(*cols_or_fields, **field_config)`

a Row can contains many Col, or the name of the fields. the related Col will be created to give equal size of all fields. the desired class can be given with a dict like construction. where the name of the field is the key, and the class is the value

the followings lines are equivalent :

```python
Row("message", "subject")
```
```python
Row(message="col-md-6", subject="col-md-6")
```
```python
Row(Col("message", class_="col-md-6"), Col("subject", class_="col-md-6"))
```

a Row should always contains Col and Col can contains Row or the field itself. a Row can be used as :

```python
layout = [
    Row(subject="col-md-8 col-xs-6", password="col-md-4 col-xs-6"),
    Row("message"),
    Row(maybemissing="col-md-12"),
]
```

## Col and Row mixed up with list and tuples

the list constuction is compiled in Row/Col construction. 
the 3 following initialization will be identical.
```python
layout = (
    ("message", "subject"),
    ("mybemissing",)
) 
layout = Col(
    ("message", "subject"),
    ("mybemissing",)
)
layout = Col(
    Row("message", "subject"),
    Row("mybemissing",)
) 
```

## Col and Row recursivity
you can add row into cols and cols into Row without deep limitation 

```python
layout = Row(
    Col(Row(Col("subject")), Row(Col("message"))),
    Col("password", "maybemissing"),
)
```

# the last hope

with this logic, it can even be possible to add more layout class with Row and Col, like Tab, Fieldset or Accordion.

**Comment by onysos** [comment]
I will work on it soon. all suggestion are welcome.

**Comment by harikvpy** [comment]
This is something that I have been planning to do for at least a few months and finally today morning decided to do it. I'm not kidding I just finished a prelimiary version and was about to post here on integrating this fix with the mainline code and lo and behold there's this issue post!!

Anyway, my implementation is pretty simple. In the form you declare a class variable 'forms_layout', where you specify the arrangement of fields like below:

fields_layout = (                                                                                             
               ('first_name', 'last_name',),                                                                         
               'address',                                                                                            
               ('email', 'phone',),                                                                                  
               ) 

Change the default FormRenderer to CustomLayoutFormRenderer in your settings.py and you get the desired result. I even updated the demo code to include a demo for this.

Implementation is in my fork at https://github.com/harikvpy/django-bootstrap3

This change does not in anyway affect the existing behavior and therefore the changes are safe to be incorporated into the original FormRenderer class.

**Comment by harikvpy** [comment]
Forgot to add -- the changes are in the branch 'layout_support'. Here's the link to the changes diff:

https://github.com/harikvpy/django-bootstrap3/commit/0c211981a562c79f04d0eb34ae0692bc7447bbc1

**Comment by dyve** [comment]
Great idea guys. Just be sure to keep the logic separated.

Some other tips

- Form.get_layout() would be OK
- layout could be a simple list of lists, or a Layout class (with Layout defined in bootstrap3)
- be sure to take interaction with layout=horizontal | vertical | inline into account

**Comment by onysos** [comment]
@harikvpy It took me many month too to find time to work on it. I don't started yet to work on code so it is cool. But I ask myself if your implementation keep some of the need I had.

The most important for me is the recursivity of the layout. I need  nested row and col to do something like the folowing form
![form_crm](https://cloud.githubusercontent.com/assets/3406733/12474902/0a43265c-c01f-11e5-8c87-a0cc280e60ec.png)

the 2nd point is something I don't have spoken about : what to do of field absent of the layout but present in the form.

>  Also, all the fields to be rendered should be specified in the layout. If
>    a field is omitted, it won't be rendered resulting in form validation error
>    if the fiels is a required field.

I intent to use the Elypsis to permit a place in the layout where to add all field not manualy added in the layout. and by default, the Elypsis will be added at the end of the Layout. 

It is clear that my intention is to build a more complex structure than yours, and in some case, it is not always the good choices (even if for myself, I need this complexity). maybe your implementation should be added in the core django_bootstrap3 and mine be added to another package ? (thanks to the `form_renderers` setting)

**Comment by harikvpy** [comment]
@dyve I forgot about the option to specify different layouts -- my forms are all vertical and therefore I just built the feature for that. Will look into addressing it. I did think about using a class, but decided to make it simple, at least to start with. And thanks for the other tips.

@onysos Some serious feature requirements there! Seriously, defaulting missing fields to current behavior ought to be easy -- as I said before I wanted something simple that I can whip up in a hour or two and therefore went for the most basic functionality. Column class for each field can be extracted from a dictionary rather than being auto-computed as it is done now.

More powerful layouts driven by a class declaration needs more work however and I'm not sure if I can find the time to embark on that at the moment. But you never know -- another relatively empty free morning of inspiration might get me going again. But I like your suggestion.

**Comment by dyve** [comment]
I love the ideas. This will be quite a job though. I might do some work on it too, to get it started properly since I actually need something like this in a project I am working on. I'll update this issue if anything happens.

**Comment by onysos** [comment]
we need it in our future projects and I am working on it at work (thanks to my open source friendly boss). so I will find the time and already have a nice base for this to work. I am writing the tests today to start checking if it work. as soon as the coverage seem enouth, I will commit to my repo.

I actually added some feature from the initial analysis : 

# Layout 

I will add a Layout class that will allow a layout by device size. by default, it will create the layout for the `FormRenderer.size`. 

# Fieldset or Accordion or Tab
it seem a nice idea to accept dict in the layout in the same way we accept list/tuple. but I don't know what this should create. a fieldset or somthing more advanced like a Tab or a Accordion/Collapse.

what I am sure is that I am writing my class taking care of the future Layout Class like them.

with this, I should be able to remove all my `{% bootstrap_field %}` from my previous projects.

**Comment by harikvpy** [comment]
@onysos I just fixed my fork to include some of the suggestions from both of you:

1. Support for form.get_layout() method, if specified. If the method is absent, class variable is checked for.
2. Custom layout based rendering only works for vertical layout. For other layouts, a message is printed to the console and the renderer falls back to default handling
3. Your ellipsis request that allows for vertical, row-by-row, arrangement for the rest of the form fields. This I believe is a good idea. If not specified, rendering stops there. I have updated the demo project to show 1 & 3 at work.

Golden rule of design is if you have multiple choices that can achieve similar results, you default to one and allow user to change it through some means -- in this case another key/option in your dict. Remember the audience for these are programmers -- some some amount of complexity is acceptable. Good luck!

**Comment by onysos** [comment]
I just have push my work. for now, the full class based layout is available and the FormRenderer is working. 34 tests was written to cover the Layout interaction. the tests currently miss the FormRenderer, but I have added it to the demo app, and it work as expected.

commit is [here](https://github.com/onysos/django-bootstrap3/commit/bb7d700edcc8d9335f943141e560ae84625f8ad6)

missing feature : 

- Elypsis managment
- retrocompatibility with FormRenderer behaviour tests
- documentation and review

I will work a it monday. 

### preview : 
the form 
```python

class LayoutContactForm(TestForm):
    """
    a form with a beautiful layout
    """
    fields_layout = [
        ("date", "dontexists", "datetime"), # row with equivalent size col. the absent field will have his space reserved
        ( # nested row and col
            ("password", "sender"),
            ("message",),
        ),
        ("cc_myself", "secret", "select1", "select2"), # row with equial size, but one is hidden. his space will be reserved
        Row("select3", Col("select4", size=8)), # mixed base and native types
        "category1", # no row nor col. just the field as by default
        # size given in a keyword fashion
        # NOTE : the order of the keywords can't be keept
        #        it will will be rendered with random order
        Row(category2=2, category3=4, category4=4),
        # full native layout
        WellLayoutElement(Row(Col(FieldContainer("addon")), size=4)) # added a custom LayoutElement
    ]

```

give you 
![form_crm1](https://cloud.githubusercontent.com/assets/3406733/12515513/169db5d8-c129-11e5-8b2b-486a5e456775.png)

**Comment by onysos** [comment]
I have made my pull request last week. did you found the time to check it ? it is mostly test covered (about 95%) and all tests passed. My LayoutFormRenderer can be a drop'in replacement of the FormRenderer by overriding the settings. nothing more is needed.

**Comment by dyve** [comment]
Sorry, no chance to look at it for now, extremely busy with work.
May take some time.

**Comment by felixcheruiyot** [comment]
Any update on this?

**Comment by harikvpy** [comment]
Short answer:
I haven't done much work on it since last update. Either reviewing the code, (sorry @onysos) or trying to improve it.

Long answer:
A couple of months back I discovered [django-crispy-forms](http://django-crispy-forms.readthedocs.io/en/latest/), which I think provides a neat interface to bootstrap forms. Essentially, it allows managing the entire layout from code, including inserting custom html (from Python code). Of course it still means layout for each form has to be composed, if the standard layout is not sufficient, but it's not that hard with the given Row() & Div() classes. Therefore I stopped using django-bootstrap3, though I'm still maintaining the existing code with the simple layout mechanism that I had built.

HTH.
